Mathlib/Data/List/Basic.lean

```diff
@@ -2559,6 +2567,85 @@ theorem foldlRecOn_nil {C : β → Sort*} (op : β → α → β) (b) (hb : C b)
   rfl
 #align list.foldl_rec_on_nil List.foldlRecOn_nil
 
+private lemma append_singleton_append_mem_left_of_length_lt {x₁ x₂ z₁ z₂ : List α} {Y₁ Y₂ : α}
+    (together : x₁ ++ [Y₁] ++ z₁ = x₂ ++ [Y₂] ++ z₂) (longer : x₂.length < x₁.length) :
+    Y₂ ∈ x₁ := by
+  have middle := congr_fun (congr_arg get? together) x₂.length
+  rw [
+    append_assoc x₂,
+    get?_append_right x₂.length.le_refl,
+    Nat.sub_self,
+    singleton_append,
+    get?_cons_zero,
+    append_assoc x₁,
+    get?_append longer,
+  ] at middle
+  exact get?_mem middle
+
+private lemma lt_iff_gt_of_add_eq_add
+    {M : Type*} [LinearOrderedAddCommMonoid M] [CovariantClass M M (· + ·) (· < ·)]
+    {a b c d : M} (total : a + b = c + d) :
+    a < c ↔ d < b := by
+  constructor
+  · intro hac
+    by_contra contr
+    rw [not_lt] at contr
+    exact (add_lt_add_of_lt_of_le hac contr).ne total
+  · intro hbd
+    by_contra contr
+    rw [not_lt] at contr
+    rw [add_comm a b, add_comm c d] at total
+    exact (add_lt_add_of_lt_of_le hbd contr).ne total.symm
+
+/- Also works:
+private lemma lt_iff_gt_of_add_eq_add' {M : Type*} [LinearOrderedSemiring M]
+    {a b c d : M} (total : a + b = c + d) :
+    a < c ↔ d < b :=
+  lt_iff_gt_of_add_eq_add total-/
+
+lemma append_singleton_append_inj_of_nmem {x₁ x₂ z₁ z₂ : List α} {Y₁ Y₂ : α}
+    (notin_x : Y₂ ∉ x₁) (notin_z : Y₂ ∉ z₁) :
+    x₁ ++ [Y₁] ++ z₁ = x₂ ++ [Y₂] ++ z₂ ↔ x₁ = x₂ ∧ Y₁ = Y₂ ∧ z₁ = z₂ := by
```

Comment:
It's quite confusing that the collections are lower-cased and the element upper-cased. If anything, I would have expected the exact opposite!

Reply:
What naming do you suggest?

Reply:
Something like this, maybe
```suggestion
lemma append_singleton_append_inj_of_nmem {x₁ x₂ z₁ z₂ : List α} {a₁ a₂ : α}
    (notin_x : a₂ ∉ x₁) (notin_z : a₂ ∉ z₁) :
    x₁ ++ [a₁] ++ z₁ = x₂ ++ [a₂] ++ z₂ ↔ x₁ = x₂ ∧ a₁ = a₂ ∧ z₁ = z₂ := by
```